### PR TITLE
chore: set nft as channel picture

### LINKED ISSUE #22: Ability to set NFT as Channel Avatar
- Click the channel picture on the settings page (/settings) should open a modal with two tabs,
  - Choose from the computer
  - Choose from the collection
- Ability to crop the image after selecting.
- Differentiate NFT channel pictures from normal ones. (channel.picture?.__typename === 'NftImage')

### PR DESCRIPTION
**Pull request by RamK777-stack**
## Description
Ability to set NFT as Channel Avatar

Closes #22

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings

### CODE REVIEW

File: apps/web/package.json [hunk 1]
```diff
@@ -45,6 +45,7 @@
     "uuid": "^9.0.0",
     "wagmi": "^0.10.2",
     "zod": "^3.19.1",
+    "react-image-crop": "^10.0.9"
```
Comment (sasicodes) [inline apps/web/package.json:48]:
```suggestion
    "react-image-crop": "^10.0.9",
```

File: apps/web/src/components/Settings/CropImageModal.tsx [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+      <div className="overflow-y-auto no-scrollbar text-center mt-2">
+        <ReactCrop
+          crop={crop}
+          onChange={(c) => setCrop(c)}
+          aspect={1}
+          className="rounded-lg"
+        >
+          <img
+            ref={imgRef}
+            src={imageSrc}
```
[the remark this answers is not in the document]
Comment (sasicodes) [reply apps/web/src/components/Settings/CropImageModal.tsx:76]: this needs to be addressed 🙇🏻

File: apps/web/src/components/Settings/CropImageModal.tsx [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+            alt="profile image"
+            className="h-80 object-fit"
+          ></img>
```
Comment (sasicodes) [inline apps/web/src/components/Settings/CropImageModal.tsx:79]: can be self-closing tag

File: packages/utils/functions/sanitizeIpfsUrl.ts [hunk 4]
```diff
@@ -6,8 +6,8 @@ const sanitizeIpfsUrl = (url: string) => {
 
   return url
     .replace(/^Qm[1-9A-Za-z]{44}/gm, `${gateway}/${url}`)
-    .replace('https://ipfs.io/ipfs', gateway)
-    .replace('https://ipfs.infura.io/ipfs', gateway)
+    .replace('https://ipfs.io/ipfs/', gateway)
```
Comment (sasicodes) [inline packages/utils/functions/sanitizeIpfsUrl.ts:9]: if we are adding `/` here, we need to remove it in constants.ts

File: apps/web/src/components/Settings/CropImageModal.tsx [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+      )
+
+      canvas.toBlob((blob) => {
+        const file = new File([blob as Blob], 'fileName.jpg', {
+          type: 'image/jpeg'
+        })
+        onPfpUpload(file)
+      }, 'image/jpeg')
+    } catch (e) {
+      console.log('crop the image')
```
Comment (sasicodes) [inline apps/web/src/components/Settings/CropImageModal.tsx:55]: add logger!

File: apps/web/src/components/Settings/ChannelPicture.tsx [hunk 6]
```diff
@@ -117,6 +124,10 @@ const ChannelPicture: FC<Props> = ({ channel }) => {
     })
   }
 
+  const [loadChallenge] = useNftChallengeLazyQuery()
+
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChannelPicture.tsx:128]:
```suggestion
```

**Review by sasicodes — CHANGES_REQUESTED** [review]

**Comment by sasicodes** [comment]
image

can we add some margin to right and bottom @RamK777-stack

File: apps/web/src/components/Settings/CropImageModal.tsx (hunk 3 above)
Comment (RamK777-stack) [reply apps/web/src/components/Settings/CropImageModal.tsx]: Resolved

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+                onChange={onChooseImage}
+              />
+            </Tab.Panel>
+            <Tab.Panel className="no-scrollbar max-h-96 overflow-y-auto focus:outline-none py-5">
+              {!error && !loading && (
+                <>
+                  <div className="grid grid-cols-3 gap-5">
+                    {collectedNFTs.map((nft: Nft) => {
+                      return (
+                        <img
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:121]: add fixed height for images.

image

**Review by sasicodes — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+type Props = {
+  isModalOpen: boolean
+  onClose?: () => void
+  onChooseImage?: (e: ChangeEvent<HTMLInputElement>) => void
+  channel: Profile
+  setNFTAvatar: (
+    contractAddress: string,
+    tokenId: string,
+    chainId: Number
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:24]:
```suggestion
    chainId: number
```

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+                  </div>
+                  {pageInfo?.next &&
+                    collectedNFTs.length !== pageInfo?.totalCount && (
+                      <span ref={observe} className="flex justify-center mt-5">
+                        <Loader />
+                      </span>
+                    )}
+                </>
+              )}
+              {loading && <TimelineShimmer />}
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:152]: move the shimmer above the map nfts and timeline shimmer looks good for a square image grid. we can use just middle spinner if needed

File: apps/web/src/components/Settings/CropImageModal.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
+import 'react-image-crop/dist/ReactCrop.css'
+
+import Modal from '@components/UIElements/Modal'
+import type { FC } from 'react'
+import React, { useRef, useState } from 'react'
+
+type Props = {
```
Comment (sasicodes) [inline apps/web/src/components/Settings/CropImageModal.tsx:7]: Move type declaration below all imports

File: apps/web/src/components/Settings/CropImageModal.tsx [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        cropped.width * scaleX,
+        cropped.height * scaleY,
+        0,
+        0,
+        cropped.width,
+        cropped.height
+      )
+
+      canvas.toBlob((blob) => {
+        const file = new File([blob as Blob], 'fileName.jpg', {
```
Comment (sasicodes) [inline apps/web/src/components/Settings/CropImageModal.tsx:50]:
```suggestion
        const file = new File([blob as Blob], 'nft.jpg', {
```

**Review by sasicodes — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+                      return (
+                        <img
+                          key={nft.contentURI}
+                          src={imageCdn(
+                            nft.originalContent?.uri
+                              ? sanitizeIpfsUrl(nft.originalContent?.uri)
+                              : `${STATIC_ASSETS}/images/placeholder.webp`,
+                            'thumbnail'
+                          )}
+                          className="rounded-xl cursor-pointer h-20 w-36"
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:131]: we can have 1:1 square image

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+                  accept=".png, .jpg, .jpeg, .svg, .gif"
+                  className="hidden w-full"
+                  onChange={onChooseImage}
+                />
+              </div>
+            </Tab.Panel>
+            <Tab.Panel className="no-scrollbar h-96 overflow-y-auto focus:outline-none py-5">
+              {!error && !loading && (
+                <>
+                  <div className="grid grid-cols-3 gap-5">
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:120]: check responsiveness in mobile view

**Review by sasicodes — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Settings/CropImageModal.tsx [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+      panelClassName="max-w-[80vh] h-auto"
+      autoClose={false}
+      title="Crop picture"
+    >
+      <div className="overflow-y-auto no-scrollbar text-center mt-2">
+        <ReactCrop
+          crop={crop}
+          onChange={(c) => setCrop(c)}
+          aspect={1}
+          className="rounded-lg"
```
Comment (sasicodes) [inline apps/web/src/components/Settings/CropImageModal.tsx:73]:
```suggestion
```

**Review by sasicodes — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 15]
```diff
@@ -0,0 +1,172 @@
...
+  return (
+    <Modal
+      onClose={() => onClose?.()}
+      show={isModalOpen}
+      panelClassName="max-w-lg h-auto no-scrollbar"
+      autoClose={false}
+      title={showCropImagePreview && 'Crop picture'}
+    >
+      {!showCropImagePreview ? (
+        <Tab.Group>
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:80]: can we make this tabgroup a separate component?

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+          <Tab.List className="flex justify-center static top-0 w-full">
+            <Tab
+              className={({ selected }) =>
+                clsx(
+                  'py-2 border-b-2 text-sm focus:outline-none w-full',
+                  selected
+                    ? 'border-indigo-500 opacity-100'
+                    : 'border-transparent opacity-50 hover:bg-indigo-500/[0.12] rounded-tl-lg rounded-tr-lg'
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:88]:
```suggestion
                    : 'border-transparent opacity-50 hover:bg-indigo-200 dark:hover:bg-indigo-800 rounded-tl-lg rounded-tr-lg'
```

File: apps/web/src/components/Settings/ChoosePicture.tsx [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+            >
+              Upload
+            </Tab>
+            <Tab
+              className={({ selected }) =>
+                clsx(
+                  'py-2 border-b-2 text-sm focus:outline-none w-full',
+                  selected
+                    ? 'border-indigo-500 opacity-100'
+                    : 'border-transparent opacity-50 hover:bg-indigo-500/[0.12] rounded-tl-lg rounded-tr-lg'
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChoosePicture.tsx:100]:
```suggestion
                    : 'border-transparent opacity-50 hover:bg-indigo-200 dark:hover:bg-indigo-800 rounded-tl-lg rounded-tr-lg'
```

File: apps/web/src/components/Settings/CropImagePreview.tsx [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  return (
+    <div className="overflow-y-auto no-scrollbar text-center mt-2">
+      <ReactCrop crop={crop} onChange={(c) => setCrop(c)} aspect={1}>
+        <img
+          ref={imgRef}
+          src={getPreviewImageSrc()}
+          alt="profile image"
+          className="h-[50vh] object-fit"
```
Comment (sasicodes) [inline apps/web/src/components/Settings/CropImagePreview.tsx:59]: this `h-50vh` causes stretch.
image

**Review by sasicodes — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Settings/ChannelPicture.tsx [hunk 19]
```diff
@@ -36,6 +39,9 @@ const ChannelPicture: FC<Props> = ({ channel }) => {
   const setSelectedChannel = useAppStore((state) => state.setSelectedChannel)
   const userSigNonce = useAppStore((state) => state.userSigNonce)
   const setUserSigNonce = useAppStore((state) => state.setUserSigNonce)
+  const [showProfileModal, setProfileModal] = useState(false)
+  const [showCropImagePreview, setShowCropImagePreview] = useState(false)
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChannelPicture.tsx:43]: do we need this state here? closing profile modal closes everything right?

**Review by sasicodes — CHANGES_REQUESTED** [review]

File: apps/web/src/components/Settings/ChannelPicture.tsx [hunk 20]
```diff
@@ -36,6 +39,9 @@ const ChannelPicture: FC<Props> = ({ channel }) => {
   const setSelectedChannel = useAppStore((state) => state.setSelectedChannel)
   const userSigNonce = useAppStore((state) => state.userSigNonce)
   const setUserSigNonce = useAppStore((state) => state.setUserSigNonce)
+  const [showProfileModal, setProfileModal] = useState(false)
```
Comment (sasicodes) [inline apps/web/src/components/Settings/ChannelPicture.tsx:42]:
```suggestion
  const [showProfileModal, setShowProfileModal] = useState(false)
```

**Review by sasicodes — CHANGES_REQUESTED** [review]

**Comment by sasicodes** [comment]
@RamK777-stack everything looks good! We need the third point to be done to differentiate nft pfp.

image

**Comment by sasicodes** [comment]
Bounty $100 - https://polygonscan.com/tx/0x192aa326a5ca221971aafd94781f819afefa034b93b5aac5b88426fcaf448df8

**Comment by sasicodes** [comment]
@RamK777-stack Any updates on this ser 🙇🏻 ?

**Comment by abhiram11** [comment]
Hi, I will try to work on this, give me some time to research and if I can do some workaround, I will request you to assign me this!

**Comment by sasicodes** [comment]
> Hi, I will try to work on this, give me some time to research and if I can do some workaround, I will request you to assign me this!

I think @RamK777-stack is yet to finish. 🙇🏻

**Comment by abhiram11** [comment]
> > Hi, I will try to work on this, give me some time to research and if I can do some workaround, I will request you to assign me this!
> 
> I think @RamK777-stack is yet to finish. 🙇🏻

Sure thing! Not in any rush, happy to assist!

**Comment by sasicodes** [comment]
> > > Hi, I will try to work on this, give me some time to research and if I can do some workaround, I will request you to assign me this!
> > 
> > 
> > I think @RamK777-stack is yet to finish. 🙇🏻
> 
> Sure thing! Not in any rush, happy to assist!

Join our discord to find the `builder` channel, where we can discuss on potential features and issues.

**Comment by sasicodes** [comment]
@RamK777-stack Please resolve conflicts 🙇🏻

**Comment by RamK777-stack** [comment]
Okay I am working on it

**Comment by sasicodes** [comment]
We can reopen once done.

**Pull request closed without merging**